In [1]:
%matplotlib inline
import matplotlib
import seaborn as sns
matplotlib.rcParams['savefig.dpi'] = 144

In [2]:
from static_grader import grader

# DW Miniproject
## Introduction

The objective of this miniproject is to exercise your ability to wrangle tabular data set and aggregate large data sets into meaningful summary statistics. We will be working with the same medical data used in the `pw` miniproject, but will be leveraging the power of Pandas to more efficiently represent and act on our data.

## Downloading the data

We first need to download the data we'll be using from Amazon S3:

In [81]:
!mkdir dw-data
!aws s3 sync s3://dataincubator-wqu/dwdata-ease/ ./dw-data

mkdir: cannot create directory 'dw-data': File exists


## Loading the data

Similar to the `PW` miniproject, the first step is to read in the data. The data files are stored as compressed CSV files. You can load the data into a Pandas DataFrame by making use of the `gzip` package to decompress the files and Panda's `read_csv` methods to parse the data into a DataFrame. You may want to check the Pandas documentation for parsing [CSV](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) files for reference.

For a description of the data set please, refer to the [PW miniproject](./pw.ipynb).

In [3]:
import pandas as pd
import numpy as np
import gzip

In [4]:
scripts_file = gzip.open('./dw-data/201701scripts_sample.csv.gz', 'r')
    
scripts = pd.read_csv(scripts_file)
scripts.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  
0     36.06        60  
1     12.49       140  
2    136.65     15000  
3     30.77         2  
4      3.36        28

In [5]:
scripts_file_16 = gzip.open('./dw-data/201606scripts_sample.csv.gz', 'r')
    
scripts_16 = pd.read_csv(scripts_file_16)
scripts_16.head()

Unnamed: 0 practice   bnf_code                             bnf_name  items  \
0           0   P84046  0101010R0            Simeticone_Dps 21mg/2.5ml      1   
1           1   N81623  040201030               Risperidone_Tab 500mcg      1   
2           2   E85658  0407020A0  Fentanyl_Transdermal Patch 12mcg/hr      2   
3           3   E82106  212300001              Sylk Vag Moist 40g Tube      2   
4           4   L83015  0403010N0              Imipramine HCl_Tab 10mg      1   

     nic  act_cost  quantity  
0   3.46      3.22       200  
1   1.15      1.18        28  
2  50.36     46.65        20  
3  10.32      9.58         2  
4   2.10      1.96        56

In [6]:
practices_file = gzip.open('./dw-data/practices.csv.gz', 'r')

col_names=[ 'code', 'name', 'addr_1', 'addr_2', 'borough', 'village', 'post_code']
practices = pd.read_csv(practices_file, names = col_names)
practices.head()

code                        name                   addr_1         addr_2  \
0  A81001         THE DENSHAM SURGERY        THE HEALTH CENTRE  LAWSON STREET   
1  A81002  QUEENS PARK MEDICAL CENTRE  QUEENS PARK MEDICAL CTR  FARRER STREET   
2  A81003   VICTORIA MEDICAL PRACTICE        THE HEALTH CENTRE  VICTORIA ROAD   
3  A81004      WOODLANDS ROAD SURGERY         6 WOODLANDS ROAD            NaN   
4  A81005          SPRINGWOOD SURGERY       SPRINGWOOD SURGERY   RECTORY LANE   

            borough    village post_code  
0  STOCKTON ON TEES  CLEVELAND  TS18 1HU  
1  STOCKTON ON TEES  CLEVELAND  TS18 2AW  
2        HARTLEPOOL  CLEVELAND  TS26 8DB  
3     MIDDLESBROUGH  CLEVELAND   TS1 3BE  
4       GUISBOROUGH        NaN  TS14 7DJ

In [7]:
chem = pd.read_csv(gzip.open('./dw-data/chem.csv.gz', 'r'))
chem.head()

CHEM SUB                                NAME
0  0101010A0                     Alexitol Sodium
1  0101010B0                          Almasilate
2  0101010C0                 Aluminium Hydroxide
3  0101010D0  Aluminium Hydroxide With Magnesium
4  0101010E0                        Hydrotalcite

now that we've loaded in the data, let's first replicate our results from the `PW` miniproject. Note that we are now working with a larger data set so the answers will be different than in the `PW` miniproject even if the analysis is the same.

## Question 1: summary_statistics

In the `PW` miniproject we first calculated the total, mean, standard deviation, and quartile statistics of the `'items'`, `'quantity'`', `'nic'`, and `'act_cost'` fields. To do this we had to write some functions to calculate the statistics and apply the functions to our data structure. The DataFrame has a `describe` method that will calculate most (not all) of these things for us.

Submit the summary statistics to the grader as a list of tuples: [('act_cost', (total, mean, std, q25, median, q75)), ...]

In [7]:
scripts.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  
0     36.06        60  
1     12.49       140  
2    136.65     15000  
3     30.77         2  
4      3.36        28

In [10]:
scripts_sum = scripts[['items','nic','act_cost','quantity']].agg(['sum'])
scripts_sum

items          nic     act_cost   quantity
sum  8982877  72490066.94  67440722.67  725345509

In [12]:
scripts_des = scripts.describe()
scripts_des

items           nic      act_cost      quantity
count  1.001634e+06  1.001634e+06  1.001634e+06  1.001634e+06
mean   8.968223e+00  7.237181e+01  6.733070e+01  7.241622e+02
std    2.966303e+01  1.913729e+02  1.774985e+02  3.882238e+03
min    1.000000e+00  0.000000e+00  1.000000e-02  0.000000e+00
25%    1.000000e+00  7.840000e+00  7.350000e+00  2.800000e+01
50%    2.000000e+00  2.256000e+01  2.113000e+01  1.000000e+02
75%    6.000000e+00  6.440000e+01  5.993000e+01  3.360000e+02
max    3.333000e+03  1.963500e+04  1.817733e+04  6.526240e+05

In [17]:
scripts_output = pd.concat([scripts_sum, scripts_des.drop(['count','min','max'])])
scripts_output

items           nic      act_cost      quantity
sum   8.982877e+06  7.249007e+07  6.744072e+07  7.253455e+08
mean  8.968223e+00  7.237181e+01  6.733070e+01  7.241622e+02
std   2.966303e+01  1.913729e+02  1.774985e+02  3.882238e+03
25%   1.000000e+00  7.840000e+00  7.350000e+00  2.800000e+01
50%   2.000000e+00  2.256000e+01  2.113000e+01  1.000000e+02
75%   6.000000e+00  6.440000e+01  5.993000e+01  3.360000e+02

In [20]:
act_cost_l = []
for i in scripts_output['act_cost']:
    act_cost_l.append(i)
    
nic_l = []
for i in scripts_output['nic']:
    nic_l.append(i)

items_l = []
for i in scripts_output['items']:
    items_l.append(i)
    
quantity_l = []
for i in scripts_output['quantity']:
    quantity_l.append(i)

#print(output_list_of_tuples)   
output_list = [('items', tuple(items_l)), ('quantity', tuple(quantity_l)), ('nic', tuple(nic_l)),('act_cost', tuple(act_cost_l))]
    

In [21]:
def summary_stats():
    return output_list
    #return [('items', (1,) * 6), ('quantity', (1,) * 6), ('nic', (1,) * 6), ('act_cost', (1,) * 6)]

In [22]:
grader.score('dw__summary_statistics', summary_stats)

Your score:  1.0


## Question 2: most_common_item

We can also easily compute summary statistics on groups within the data. In the `pw` miniproject we had to explicitly construct the groups based on the values of a particular field. Pandas will handle that for us via the `groupby` method. This process is [detailed in the Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/groupby.html).

Use `groupby` to calculate the total number of items dispensed for each `'bnf_name'`. Find the item with the highest total and return the result as `(bnf_name, total)`.

In [9]:
scripts.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  
0     36.06        60  
1     12.49       140  
2    136.65     15000  
3     30.77         2  
4      3.36        28

In [15]:
scripts_gp = scripts.groupby('bnf_name')['items'].count()
scripts_gp

bnf_name
365 Film 4cm x 5cm VP Adh Film Dress          1
365 IV Transpt IV 10cm x 12cm VP Adh Fil      1
365 Non Adherent 10cm x 10cm Pfa Plas Fa      1
365 Non Woven Island 10cm x 10cm Adh Dre      2
3M Micropore Silicone 2.5cm x 5m Surg Ad     14
3M Micropore Silicone 5cm x 5m Surg Adh       7
3m Health Care_Cavilon Durable Barrier C    913
3m Health Care_Cavilon No Sting 1ml Barr    209
3m Health Care_Cavilon No Sting 3ml Barr     88
3m Health Care_Cavilon No Sting Barrier     466
4Head_Top Headache Relief Stick               4
A & P_Infants Pdrs                            4
A.S Saliva Orthana Spy 50ml (App)           116
A.S Saliva Orthana Spy Refill 500ml (App      6
A2A Spacer                                   30
A2A Spacer + Sml/Med Mask                    73
AAA_Sore Throat A/Spy 7.5g                    3
AMI_Corsinel Suportx Ab Tube Wte ExLge (      1
AMI_Corsinel Suportx Ab Tube Wte Lge (94      1
AMI_Corsinel Suportx Ab Tube Wte Med (86      2
AMI_Corsinel Suportx Easy Panel

In [22]:
sums = scripts.groupby('bnf_name').sum()['items']
sums.head()

bnf_name
365 Film 4cm x 5cm VP Adh Film Dress         1
365 IV Transpt IV 10cm x 12cm VP Adh Fil     1
365 Non Adherent 10cm x 10cm Pfa Plas Fa     1
365 Non Woven Island 10cm x 10cm Adh Dre     4
3M Micropore Silicone 2.5cm x 5m Surg Ad    16
Name: items, dtype: int64

In [16]:
def most_common_item():
    sums = scripts.groupby('bnf_name').sum()['items']
    item = (sums.idxmax(), sums.max())
    return [item]

In [18]:
answer = most_common_item()
answer

[('Omeprazole_Cap E/C 20mg', 222007)]

In [54]:
def most_common_item():
    return (0, 1643)

In [17]:
grader.score('dw__most_common_item', most_common_item)

Your score:  1.0


## Question 3: items_by_region

Now let's find the most common item by post code. The post code information is in the `practices` DataFrame, and we'll need to `merge` it into the `scripts` DataFrame. Pandas provides [extensive documentation](https://pandas.pydata.org/pandas-docs/stable/merging.html) with diagrammed examples on different methods and approaches for joining data. The `merge` method is only one of many possible options.

Return your results as a list of tuples `(post code, item name, amount dispensed as % of total)`. Sort your results ascending alphabetically by post code and take only results from the first 100 post codes.

**NOTE:** Some practices have multiple postal codes associated with them. Use the alphabetically first postal code. Note some postal codes may have multiple `'bnf_name'` with the same prescription rate for the maximum. In this case, take the alphabetically first `'bnf_name'` (as in the PW miniproject).

In [8]:
practices.head()

code                        name                   addr_1         addr_2  \
0  A81001         THE DENSHAM SURGERY        THE HEALTH CENTRE  LAWSON STREET   
1  A81002  QUEENS PARK MEDICAL CENTRE  QUEENS PARK MEDICAL CTR  FARRER STREET   
2  A81003   VICTORIA MEDICAL PRACTICE        THE HEALTH CENTRE  VICTORIA ROAD   
3  A81004      WOODLANDS ROAD SURGERY         6 WOODLANDS ROAD            NaN   
4  A81005          SPRINGWOOD SURGERY       SPRINGWOOD SURGERY   RECTORY LANE   

            borough    village post_code  
0  STOCKTON ON TEES  CLEVELAND  TS18 1HU  
1  STOCKTON ON TEES  CLEVELAND  TS18 2AW  
2        HARTLEPOOL  CLEVELAND  TS26 8DB  
3     MIDDLESBROUGH  CLEVELAND   TS1 3BE  
4       GUISBOROUGH        NaN  TS14 7DJ

In [9]:
scripts.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  
0     36.06        60  
1     12.49       140  
2    136.65     15000  
3     30.77         2  
4      3.36        28

In [10]:
srt = practices.sort_values('post_code').drop_duplicates('code', keep='first')
srt.head()

code                               name                   addr_1  \
1896  E82060             PARKBURY HOUSE SURGERY           PARKBURY HOUSE   
1871  E82031                   MALTINGS SURGERY     THE MALTINGS SURGERY   
1849  E82004              HATFIELD ROAD SURGERY         61 HATFIELD ROAD   
1848  E82002              WRAFTON HOUSE SURGERY    WRAFTON HOUSE SURGERY   
9812  Y05146  HCT LYMPHOEDEMA AT WEST ESSEX CCG  QUEENSWAY HEALTH CENTRE   

                    addr_2    borough        village post_code  
1896      ST.PETERS STREET  ST.ALBANS  HERTFORDSHIRE   AL1 3HD  
1871  8-14 VICTORIA STREET  ST ALBANS  HERTFORDSHIRE   AL1 3JB  
1849                   NaN  ST.ALBANS  HERTFORDSHIRE   AL1 4JE  
1848   9/11 WELLFIELD ROAD   HATFIELD  HERTFORDSHIRE  AL10 0BS  
9812             QUEENSWAY   HATFIELD  HERTFORDSHIRE  AL10 0LF

In [11]:
merged = scripts.merge(srt, left_on='practice', right_on='code')
merged.head()

practice   bnf_code                                  bnf_name  items  \
0   B87016  213200001       MucoClear Sod Chlor 6% Inh Soln 4ml      1   
1   B87016  0704020N0                        Detrusitol_Tab 2mg      2   
2   B87016  0604011G0                      Elleste Solo_Tab 2mg      7   
3   B87016  0407010F0             Co-Codamol Eff_Tab 30mg/500mg     16   
4   B87016  091000000  Centrum Advance 50+_Multivit/Mineral Tab      1   

      nic  act_cost  quantity    code                name              addr_1  \
0   38.94     36.06        60  B87016  WHITE ROSE SURGERY  WHITE ROSE SURGERY   
1   61.12     56.60       112  B87016  WHITE ROSE SURGERY  WHITE ROSE SURGERY   
2   35.42     32.89       588  B87016  WHITE ROSE SURGERY  WHITE ROSE SURGERY   
3  169.54    158.06      1958  B87016  WHITE ROSE SURGERY  WHITE ROSE SURGERY   
4    3.66      3.40        30  B87016  WHITE ROSE SURGERY  WHITE ROSE SURGERY   

                      addr_2     borough         village post_code  
0  EXCHANGE ST SOUTH ELMSALL  PONTEFRACT  WEST YORKSHIRE   WF9 2RD  
1  EXCHANGE ST SOUTH ELMSALL  PONTEFRACT  WEST YORKSHIRE   WF9 2RD  
2  EXCHANGE ST SOUTH ELMSALL  PONTEFRACT  WEST YORKSHIRE   WF9 2RD  
3  EXCHANGE ST SOUTH ELMSALL  PONTEFRACT  WEST YORKSHIRE   WF9 2RD  
4  EXCHANGE ST SOUTH ELMSALL  PONTEFRACT  WEST YORKSHIRE   WF9 2RD

In [12]:
total_items_by_postcode = merged.groupby(['post_code','bnf_name'])[['items']].sum()
total_items_by_postcode.head()

items
post_code bnf_name                                       
AL1 3HD   Acetic Acid_Ear Spy 2% 5ml                    2
          ActivHeal 10cm x 10cm Non-Adh Foam Wound      1
          Adcal-D3_Capl 750mg/200u                     35
          AgaMatrix Ultra-Thin Lancets 0.35mm/28 G      5
          Amiodarone HCl_Tab 200mg                      5

In [13]:
total_items_by_postcode.reset_index(inplace=True)
total_items_by_postcode.head()

post_code                                  bnf_name  items
0   AL1 3HD                Acetic Acid_Ear Spy 2% 5ml      2
1   AL1 3HD  ActivHeal 10cm x 10cm Non-Adh Foam Wound      1
2   AL1 3HD                  Adcal-D3_Capl 750mg/200u     35
3   AL1 3HD  AgaMatrix Ultra-Thin Lancets 0.35mm/28 G      5
4   AL1 3HD                  Amiodarone HCl_Tab 200mg      5

In [14]:
total_items_max = total_items_by_postcode.groupby('post_code')[['items']].max()
total_items_max.head()

items
post_code       
AL1 3HD      187
AL1 3JB      225
AL1 4JE       70
AL10 0BS     180
AL10 0LF       1

In [15]:
total_items_max.reset_index(inplace=True)
total_items_max.head()

post_code  items
0   AL1 3HD    187
1   AL1 3JB    225
2   AL1 4JE     70
3  AL10 0BS    180
4  AL10 0LF      1

In [16]:
total_items = total_items_max.merge(total_items_by_postcode, on=['post_code','items'], how = 'left')

In [17]:
total_items.head()

post_code  items                       bnf_name
0   AL1 3HD    187          Amoxicillin_Cap 500mg
1   AL1 3HD    187       Levothyrox Sod_Tab 25mcg
2   AL1 3JB    225  Bendroflumethiazide_Tab 2.5mg
3   AL1 4JE     70               Aspirin_Tab 75mg
4  AL10 0BS    180          Amoxicillin_Cap 500mg

In [18]:
total_items_sorted = total_items.sort_values('bnf_name').drop_duplicates(['post_code','items'], keep='first')
total_items_sorted.head()

post_code  items                                  bnf_name
2558  GL53 9QU      4  3m Health Care_Cavilon Durable Barrier C
67     B14 7AG     53  3m Health Care_Cavilon Durable Barrier C
111    B24 9JN    162  3m Health Care_Cavilon Durable Barrier C
6439   SR4 7TP      1  3m Health Care_Cavilon No Sting 1ml Barr
7647   WN1 1NJ      2  3m Health Care_Cavilon No Sting 1ml Barr

In [19]:
total_items_sorted.sort_values('post_code', inplace=True)
total_items_sorted.head()

post_code  items                                  bnf_name
0   AL1 3HD    187                     Amoxicillin_Cap 500mg
2   AL1 3JB    225             Bendroflumethiazide_Tab 2.5mg
3   AL1 4JE     70                          Aspirin_Tab 75mg
4  AL10 0BS    180                     Amoxicillin_Cap 500mg
5  AL10 0LF      1  ActiLymph Class 1 Combined Armsleeve + T

In [35]:
total_items_sorted.shape

(7572, 3)

In [30]:
s = total_items_by_postcode.groupby('post_code')[['items']].sum().reset_index()
s.head()

post_code  items
0   AL1 3HD   1822
1   AL1 3JB   1778
2   AL1 4JE    364
3  AL10 0BS   1451
4  AL10 0LF      3

In [36]:
s.shape

(7572, 2)

In [37]:
final_df = s.merge(total_items_sorted, on='post_code')
final_df.head()

post_code  items_x  items_y                                  bnf_name
0   AL1 3HD     1822      187                     Amoxicillin_Cap 500mg
1   AL1 3JB     1778      225             Bendroflumethiazide_Tab 2.5mg
2   AL1 4JE      364       70                          Aspirin_Tab 75mg
3  AL10 0BS     1451      180                     Amoxicillin_Cap 500mg
4  AL10 0LF        3        1  ActiLymph Class 1 Combined Armsleeve + T

In [38]:
final_df['amt%'] = final_df['items_y'].astype(float)/final_df['items_x'].astype(float)
final_df['amt%'].head()

0    0.102634
1    0.126547
2    0.192308
3    0.124052
4    0.333333
Name: amt%, dtype: float64

In [39]:
final_df.head()

post_code  items_x  items_y                                  bnf_name  \
0   AL1 3HD     1822      187                     Amoxicillin_Cap 500mg   
1   AL1 3JB     1778      225             Bendroflumethiazide_Tab 2.5mg   
2   AL1 4JE      364       70                          Aspirin_Tab 75mg   
3  AL10 0BS     1451      180                     Amoxicillin_Cap 500mg   
4  AL10 0LF        3        1  ActiLymph Class 1 Combined Armsleeve + T   

       amt%  
0  0.102634  
1  0.126547  
2  0.192308  
3  0.124052  
4  0.333333

In [53]:
result = final_df[['post_code','bnf_name','amt%']]
result.head()

post_code                                  bnf_name      amt%
0   AL1 3HD                     Amoxicillin_Cap 500mg  0.102634
1   AL1 3JB             Bendroflumethiazide_Tab 2.5mg  0.126547
2   AL1 4JE                          Aspirin_Tab 75mg  0.192308
3  AL10 0BS                     Amoxicillin_Cap 500mg  0.124052
4  AL10 0LF  ActiLymph Class 1 Combined Armsleeve + T  0.333333

In [57]:
result = result.head(100)
values = result.get_values().tolist()

In [58]:
final=[]

for item in values:
    final.append(tuple(item))

In [72]:
def items_by_region():
    return final

In [73]:
print items_by_region()

[('AL1 3HD', 'Amoxicillin_Cap 500mg', 0.1026344676180022), ('AL1 3JB', 'Bendroflumethiazide_Tab 2.5mg', 0.1265466816647919), ('AL1 4JE', 'Aspirin_Tab 75mg', 0.19230769230769232), ('AL10 0BS', 'Amoxicillin_Cap 500mg', 0.12405237767057202), ('AL10 0LF', 'ActiLymph Class 1 Combined Armsleeve + T', 0.3333333333333333), ('AL10 0NL', 'Amitriptyline HCl_Tab 10mg', 0.0639686684073107), ('AL10 0UR', 'Diazepam_Tab 10mg', 0.5434782608695652), ('AL10 8HP', 'Sertraline HCl_Tab 50mg', 0.10324129651860744), ('AL2 1ES', 'Levothyrox Sod_Tab 100mcg', 0.13074204946996468), ('AL2 3JX', 'Simvastatin_Tab 40mg', 0.0847231487658439), ('AL3 5ER', 'Bisoprolol Fumar_Tab 2.5mg', 0.11428571428571428), ('AL3 5HB', 'Omeprazole_Cap E/C 20mg', 0.16846758349705304), ('AL3 5JB', 'Alimemazine Tart_Tab 10mg', 1.0), ('AL3 5NF', 'Ramipril_Cap 10mg', 0.09449465899753492), ('AL3 5NP', 'Clopidogrel_Tab 75mg', 0.09023255813953489), ('AL3 7BL', 'Bendroflumethiazide_Tab 2.5mg', 0.08917197452229299), ('AL3 8LJ', 'Aspirin Disper_Ta

In [37]:
def items_by_region():
    [('AL1 3HD', u'Amoxicillin_Cap 500mg', 0.1026344676180022)] * 100

In [61]:
grader.score('dw__items_by_region', items_by_region)

Your score:  1.0


## Question 4: script_anomalies

Drug abuse is a source of human and monetary costs in health care. A first step in identifying practitioners that enable drug abuse is to look for practices where commonly abused drugs are prescribed unusually often. Let's try to find practices that prescribe an unusually high amount of opioids. The opioids we'll look for are given in the list below.

In [8]:
opioids = ['morphine', 'oxycodone', 'methadone', 'fentanyl', 'pethidine', 'buprenorphine', 'propoxyphene', 'codeine']

In [9]:
opioids_join = '|'.join(opioids)
opioids_join

'morphine|oxycodone|methadone|fentanyl|pethidine|buprenorphine|propoxyphene|codeine'

In [10]:
chem.head()

CHEM SUB                                NAME
0  0101010A0                     Alexitol Sodium
1  0101010B0                          Almasilate
2  0101010C0                 Aluminium Hydroxide
3  0101010D0  Aluminium Hydroxide With Magnesium
4  0101010E0                        Hydrotalcite

In [11]:
opioid_codes = chem.loc[chem['NAME'].str.contains(opioids_join, case=False)]['CHEM SUB'].tolist()
len(opioid_codes)

35

In [12]:
chem_new = chem
chem_new.columns = ['bnf_code','chem_name']
chem_new.head()

bnf_code                           chem_name
0  0101010A0                     Alexitol Sodium
1  0101010B0                          Almasilate
2  0101010C0                 Aluminium Hydroxide
3  0101010D0  Aluminium Hydroxide With Magnesium
4  0101010E0                        Hydrotalcite

In [13]:
scripts.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  
0     36.06        60  
1     12.49       140  
2    136.65     15000  
3     30.77         2  
4      3.36        28

In [14]:
scripts.shape

(1001634, 7)

In [15]:
scrpt = scripts
scrpt.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  
0     36.06        60  
1     12.49       140  
2    136.65     15000  
3     30.77         2  
4      3.36        28

In [16]:
scrpt['opioid_prescription'] = scrpt['bnf_code'].isin(opioid_codes)

In [17]:
scrpt.shape

(1001634, 8)

In [18]:
scrpt.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity  opioid_prescription  
0     36.06        60                False  
1     12.49       140                False  
2    136.65     15000                False  
3     30.77         2                False  
4      3.36        28                False

In [19]:
practices.head()

code                        name                   addr_1         addr_2  \
0  A81001         THE DENSHAM SURGERY        THE HEALTH CENTRE  LAWSON STREET   
1  A81002  QUEENS PARK MEDICAL CENTRE  QUEENS PARK MEDICAL CTR  FARRER STREET   
2  A81003   VICTORIA MEDICAL PRACTICE        THE HEALTH CENTRE  VICTORIA ROAD   
3  A81004      WOODLANDS ROAD SURGERY         6 WOODLANDS ROAD            NaN   
4  A81005          SPRINGWOOD SURGERY       SPRINGWOOD SURGERY   RECTORY LANE   

            borough    village post_code  
0  STOCKTON ON TEES  CLEVELAND  TS18 1HU  
1  STOCKTON ON TEES  CLEVELAND  TS18 2AW  
2        HARTLEPOOL  CLEVELAND  TS26 8DB  
3     MIDDLESBROUGH  CLEVELAND   TS1 3BE  
4       GUISBOROUGH        NaN  TS14 7DJ

In [20]:
pract = practices[['code', 'name']]
pract.columns = ['practice', 'name']
pract.head()

practice                        name
0   A81001         THE DENSHAM SURGERY
1   A81002  QUEENS PARK MEDICAL CENTRE
2   A81003   VICTORIA MEDICAL PRACTICE
3   A81004      WOODLANDS ROAD SURGERY
4   A81005          SPRINGWOOD SURGERY

In [21]:
len(scrpt.loc[scrpt['opioid_prescription']])

34597

In [22]:
opioids_per_practice = scrpt.groupby('practice')['opioid_prescription'].mean().rename('frac')
opioids_per_practice.head()

practice
A81001    0.025424
A81002    0.021834
A81004    0.048780
A81005    0.034965
A81006    0.036269
Name: frac, dtype: float64

In [23]:
opioids_per_practice_std = scrpt.groupby('practice')['opioid_prescription'].std().rename('frac_std')
opioids_per_practice_std.head()

practice
A81001    0.158080
A81002    0.146462
A81004    0.216069
A81005    0.184337
A81006    0.187446
Name: frac_std, dtype: float64

In [24]:
overall_rate = scrpt['opioid_prescription'].mean()
overall_rate

0.034540560723777348

In [25]:
overall_rate_std = scrpt['opioid_prescription'].std()
overall_rate_std

0.18261309833034187

In [26]:
relative_opioids_per_practice = (opioids_per_practice - overall_rate).rename('relative')
relative_opioids_per_practice.head()

practice
A81001   -0.009117
A81002   -0.012706
A81004    0.014240
A81005    0.000424
A81006    0.001729
Name: relative, dtype: float64

In [27]:
opioid = scrpt.groupby('practice')['opioid_prescription'].sum().rename('opioid')
opioid.head()

practice
A81001    3.0
A81002    5.0
A81004    8.0
A81005    5.0
A81006    7.0
Name: opioid, dtype: float64

In [28]:
total = scrpt.groupby('practice')['bnf_code'].count().rename('total')
total.head()

practice
A81001    118
A81002    229
A81004    164
A81005    143
A81006    193
Name: total, dtype: int64

In [29]:
standard_error_per_practice = (overall_rate_std/(total**0.5)).rename('std_err')
standard_error_per_practice.head()

practice
A81001    0.016811
A81002    0.012067
A81004    0.014260
A81005    0.015271
A81006    0.013145
Name: std_err, dtype: float64

In [30]:
opioid_scores = (relative_opioids_per_practice/standard_error_per_practice).rename('opioid_scores')
opioid_scores.head()

practice
A81001   -0.542317
A81002   -1.052960
A81004    0.998614
A81005    0.027796
A81006    0.131525
Name: opioid_scores, dtype: float64

In [31]:
merged = pd.concat([opioid, total, opioids_per_practice, relative_opioids_per_practice, standard_error_per_practice,opioid_scores], axis=1)

In [42]:
merged = merged.reset_index()

In [43]:
merged.head()

practice  opioid  total      frac  relative   std_err  opioid_scores
0   A81001     3.0    118  0.025424 -0.009117  0.016811      -0.542317
1   A81002     5.0    229  0.021834 -0.012706  0.012067      -1.052960
2   A81004     8.0    164  0.048780  0.014240  0.014260       0.998614
3   A81005     5.0    143  0.034965  0.000424  0.015271       0.027796
4   A81006     7.0    193  0.036269  0.001729  0.013145       0.131525

In [33]:
type(merged)

pandas.core.frame.DataFrame

In [34]:
merged.shape

(9230, 6)

In [84]:
pract.reset_index().head()

index practice                        name
0      0   A81001         THE DENSHAM SURGERY
1      1   A81002  QUEENS PARK MEDICAL CENTRE
2      2   A81003   VICTORIA MEDICAL PRACTICE
3      3   A81004      WOODLANDS ROAD SURGERY
4      4   A81005          SPRINGWOOD SURGERY

In [85]:
type(pract)

pandas.core.frame.DataFrame

In [86]:
final_df = merged.merge(pract, on='practice', how='left')

In [87]:
final_df.head()

practice  opioid  total      frac  relative   std_err  opioid_scores  \
0   A81001     3.0    118  0.025424 -0.009117  0.016811      -0.542317   
1   A81001     3.0    118  0.025424 -0.009117  0.016811      -0.542317   
2   A81002     5.0    229  0.021834 -0.012706  0.012067      -1.052960   
3   A81004     8.0    164  0.048780  0.014240  0.014260       0.998614   
4   A81004     8.0    164  0.048780  0.014240  0.014260       0.998614   

                         name  
0         THE DENSHAM SURGERY  
1         THE DENSHAM SURGERY  
2  QUEENS PARK MEDICAL CENTRE  
3      WOODLANDS ROAD SURGERY  
4     BLUEBELL MEDICAL CENTRE

In [88]:
final_df.sort_values('opioid_scores', ascending = False , inplace=True)

In [89]:
final_df.head()

practice  opioid  total  frac  relative   std_err  opioid_scores  \
9644   Y04576     4.0      4   1.0  0.965459  0.091307      10.573825   
7795   P88636     4.0      4   1.0  0.965459  0.091307      10.573825   
9645   Y04576     4.0      4   1.0  0.965459  0.091307      10.573825   
8718   Y02770     4.0      5   0.8  0.765459  0.081667       9.372928   
8332   Y01640     4.0      5   0.8  0.765459  0.081667       9.372928   

                       name  
9644  ADDACTION NORTH DEVON  
7795    COMMUNITY DRUG TEAM  
9645  ADDACTION NORTH DEVON  
8718  NY HORIZONS HARROGATE  
8332      SALFORD DRUG TEAM

In [90]:
final = final_df.drop_duplicates('name')
final.head()

practice  opioid  total  frac  relative   std_err  opioid_scores  \
9644   Y04576     4.0      4   1.0  0.965459  0.091307      10.573825   
7795   P88636     4.0      4   1.0  0.965459  0.091307      10.573825   
8718   Y02770     4.0      5   0.8  0.765459  0.081667       9.372928   
8332   Y01640     4.0      5   0.8  0.765459  0.081667       9.372928   
9079   Y03738     3.0      3   1.0  0.965459  0.105432       9.157201   

                       name  
9644  ADDACTION NORTH DEVON  
7795    COMMUNITY DRUG TEAM  
8718  NY HORIZONS HARROGATE  
8332      SALFORD DRUG TEAM  
9079       CRI DRUG SERVICE

In [97]:
result = final[['name','opioid_scores','total']]
result.head()

name  opioid_scores  total
9644  ADDACTION NORTH DEVON      10.573825      4
7795    COMMUNITY DRUG TEAM      10.573825      4
8718  NY HORIZONS HARROGATE       9.372928      5
8332      SALFORD DRUG TEAM       9.372928      5
9079       CRI DRUG SERVICE       9.157201      3

In [98]:
result = result.head(100)
values = result.get_values().tolist()

In [99]:
answer=[]

for item in values:
    answer.append(tuple(item))

In [100]:
def script_anomalies():
    return answer

In [101]:
print script_anomalies()

[('ADDACTION NORTH DEVON', 10.573824639125657, 4), ('COMMUNITY DRUG TEAM', 10.573824639125657, 4), ('NY HORIZONS HARROGATE', 9.3729275495027, 5), ('SALFORD DRUG TEAM', 9.3729275495027, 5), ('CRI DRUG SERVICE', 9.157200752644645, 3), ('CRI CAMDEN COMMUNITY DRUG TREATMENT SERV', 9.157200752644645, 3), ('WORCESTERSHIRE RECOVERY PARTNERSHIP', 9.157200752644645, 3), ('HAYFIELD GPWSI SERVICE', 9.157200752644645, 3), ('ADDICTIONS SERVICE', 9.145409115671335, 8), ('TURNING POINT CROYDON RECOVERY NETWORK', 8.479054497238709, 6), ('DUDLEY COMMUNITY PALLIATIVE MEDICINE', 8.060276069984052, 10), ('MACMILLAN ST HELENS SP  PALLIATIVE CARE', 7.835795414652864, 4), ('LIFELINE SOUTHWARK', 7.835795414652864, 4), ('STAR - EAST SUSSEX DRUG & ALCOHOL SERV', 7.835795414652864, 4), ('PLYMOUTH SPECIALIST ADDICTION SERVICE', 7.835795414652864, 4), ("ST ANN'S HOSPICE/PAL CARE", 7.835795414652864, 4), ('WESTMINSTER DAWS', 7.835795414652864, 4), ('COMMUNITY/SLH DAY CASE PALLIATIVE CARE', 7.778588563496554, 7), ('

These are generic names for drugs, not brand names. Generic drug names can be found using the `'bnf_code'` field in `scripts` along with the `chem` table.. Use the list of opioids provided above along with these fields to make a new field in the `scripts` data that flags whether the row corresponds with a opioid prescription.

In [ ]:
scripts_chem = pd.merge(scripts, chem, left_on='bnf_code', right_on='CHEM SUB')
scripts_chem.head()

### Note
Owing to helpful comments and answers, I managed to partially resolve problems. Here are my intermediate results...

Number of opioids in chem dataframe is __35__. Number of opioid prescriptions in scripts dataframe is __34597__. There is __9230__ practices in total, to be analyzed. The mean value of opioid prescription rate for the whole population is 0.0345405607238. From individual practices means, I subtracted the total mean to get relative prescription rate. Standard deviation (std) for the whole population (based on binomial distribution) is 0.18261309833. For each practice I determined standard error by dividing standard deviation with the square root of total prescriptions, including opioid and non-opioid ones. Finally, I determined opioid scores by dividing relative prescription rate with standard error. Floats were used everywhere, to avoid accidental integer division.

Now for each practice calculate the proportion of its prescriptions containing opioids.

**Hint:** Consider the following list: `[0, 1, 1, 0, 0, 0]`. What proportion of the entries are 1s? What is the mean value?

In [ ]:
opioids_per_practice = ...

How do these proportions compare to the overall opioid prescription rate? Subtract off the proportion of all prescriptions that are opioids from each practice's proportion.

In [ ]:
relative_opioids_per_practice = ...

Now that we know the difference between each practice's opioid prescription rate and the overall rate, we can identify which practices prescribe opioids at above average or below average rates. However, are the differences from the overall rate important or just random deviations? In other words, are the differences from the overall rate big or small?

To answer this question we have to quantify the difference we would typically expect between a given practice's opioid prescription rate and the overall rate. This quantity is called the **standard error**, and is related to the **standard deviation**, $\sigma$. The standard error in this case is

$$ \frac{\sigma}{\sqrt{n}} $$

where $n$ is the number of prescriptions each practice made. Calculate the standard error for each practice. Then divide `relative_opioids_per_practice` by the standard errors. We'll call the final result `opioid_scores`.

In [ ]:
standard_error_per_practice = ...
opioid_scores = ...

The quantity we have calculated in `opioid_scores` is called a **z-score**:

$$ \frac{\bar{X} - \mu}{\sqrt{\sigma^2/n}} $$

Here $\bar{X}$ corresponds with the proportion for each practice, $\mu$ corresponds with the proportion across all practices, $\sigma^2$ corresponds with the variance of the proportion across all practices, and $n$ is the number of prescriptions made by each practice. Notice $\bar{X}$ and $n$ will be different for each practice, while $\mu$ and $\sigma$ are determined across all prescriptions, and so are the same for every z-score. The z-score is a useful statistical tool used for hypothesis testing, finding outliers, and comparing data about different types of objects or events.

Now that we've calculated this statistic, take the 100 practices with the largest z-score. Return your result as a list of tuples in the form `(practice_name, z-score, number_of_scripts)`. Sort your tuples by z-score in descending order. Note that some practice codes will correspond with multiple names. In this case, use the first match when sorting names alphabetically.

In [ ]:
unique_practices = ...
anomalies = ...

In [44]:
def script_anomalies():
    return [("ADDACTION NORTH DEVON", 10.5738246391, 4.0)] * 100

In [102]:
grader.score('dw__script_anomalies', script_anomalies)

Your score:  0.94


## Question 5: script_growth

Another way to identify anomalies is by comparing current data to historical data. In the case of identifying sites of drug abuse, we might compare a practice's current rate of opioid prescription to their rate 5 or 10 years ago. Unless the nature of the practice has changed, the profile of drugs they prescribe should be relatively stable. We might also want to identify trends through time for business reasons, identifying drugs that are gaining market share. That's what we'll do in this question.

We'll load in beneficiary data from 6 months earlier, June 2016, and calculate the percent growth in prescription rate from June 2016 to January 2017 for each `bnf_name`.  Normalize the percent growth in prescriptions of individual items by the percent change in total number of prescriptions (think about whether this normalization should be a division or a subtraction). We'll return the 50 items with largest growth and the 50 items with the largest shrinkage (i.e. negative percent growth) as a list of tuples sorted by growth rate in descending order in the format `(script_name, growth_rate, raw_2016_count)`. You'll notice that many of the 50 fastest growing items have low counts of prescriptions in 2016. Filter out any items that were prescribed less than 50 times.

In [206]:
drugs_16=scripts_16[['bnf_name', 'items']]
drugs_16=drugs_16.groupby('bnf_name').count().reset_index().drop_duplicates()
drugs_16.columns=[['bnf_name', 'count_16']]
drugs_16.head()

bnf_name  count_16
0      365 Film 4cm x 5cm VP Adh Film Dress         1
1  365 Non Adherent 10cm x 10cm Pfa Plas Fa         1
2  365 Non Adherent 5cm x 5cm Pfa Plas Face         2
3  365 Non Woven Island 10cm x 20cm Adh Dre         1
4  365 Non Woven Island 5cm x 7.2cm Adh Dre         1

In [207]:
drugs_17=scripts[['bnf_name', 'items']]
drugs_17=drugs_17.groupby('bnf_name').count().reset_index().drop_duplicates()
drugs_17.columns=[['bnf_name', 'count_17']]
drugs_17.head()

bnf_name  count_17
0      365 Film 4cm x 5cm VP Adh Film Dress         1
1  365 IV Transpt IV 10cm x 12cm VP Adh Fil         1
2  365 Non Adherent 10cm x 10cm Pfa Plas Fa         1
3  365 Non Woven Island 10cm x 10cm Adh Dre         2
4  3M Micropore Silicone 2.5cm x 5m Surg Ad        14

In [208]:
total_16=drugs_16['count_16'].sum()
total_17=drugs_17['count_17'].sum()

In [209]:
drugs_16['rate_16']=drugs_16['count_16']/total_16.astype(float)
drugs_17['rate_17']=drugs_17['count_17']/total_17.astype(float)

In [210]:
drugs=drugs_16.merge(drugs_17, on='bnf_name', how='inner').drop_duplicates('bnf_name')
drugs.head()

bnf_name  count_16       rate_16  count_17  \
0      365 Film 4cm x 5cm VP Adh Film Dress         1  9.765577e-07         1   
1  365 Non Adherent 10cm x 10cm Pfa Plas Fa         1  9.765577e-07         1   
2  3m Health Care_Cavilon Durable Barrier C       905  8.837847e-04       913   
3  3m Health Care_Cavilon No Sting 1ml Barr       228  2.226552e-04       209   
4  3m Health Care_Cavilon No Sting 3ml Barr       103  1.005854e-04        88   

        rate_17  
0  9.983687e-07  
1  9.983687e-07  
2  9.115106e-04  
3  2.086591e-04  
4  8.785644e-05

In [211]:
drugs['growth']=(drugs['rate_17']-drugs['rate_16'])/drugs['rate_16'].astype(float)
drugs.head()

bnf_name  count_16       rate_16  count_17  \
0      365 Film 4cm x 5cm VP Adh Film Dress         1  9.765577e-07         1   
1  365 Non Adherent 10cm x 10cm Pfa Plas Fa         1  9.765577e-07         1   
2  3m Health Care_Cavilon Durable Barrier C       905  8.837847e-04       913   
3  3m Health Care_Cavilon No Sting 1ml Barr       228  2.226552e-04       209   
4  3m Health Care_Cavilon No Sting 3ml Barr       103  1.005854e-04        88   

        rate_17    growth  
0  9.983687e-07  0.022335  
1  9.983687e-07  0.022335  
2  9.115106e-04  0.031372  
3  2.086591e-04 -0.062860  
4  8.785644e-05 -0.126549

In [212]:
total_growth=(total_17-total_16)/total_16.astype(float) #-0.0218465730148
total_growth

-0.021846573014780202

In [213]:
drugs['rel_growth']=drugs['growth']+total_growth
drugs=drugs[(drugs['count_16']>50)]
drugs = drugs[['bnf_name', 'rel_growth', 'count_16']]
drugs.head()

bnf_name  rel_growth  count_16
2  3m Health Care_Cavilon Durable Barrier C    0.009525       905
3  3m Health Care_Cavilon No Sting 1ml Barr   -0.084707       228
4  3m Health Care_Cavilon No Sting 3ml Barr   -0.148396       103
5   3m Health Care_Cavilon No Sting Barrier   -0.152489       548
8         A.S Saliva Orthana Spy 50ml (App)    0.176040        99

In [214]:
drugs.sort_values('rel_growth', ascending=False, inplace=True)
drugs.head()

bnf_name  rel_growth  count_16
1483          Butec_Transdermal Patch 10mcg/hr    3.946340        57
1485           Butec_Transdermal Patch 5mcg/hr    2.672189        75
4483  Fostair NEXThaler_Inh 200mcg/6mcg (120D)    1.726507        77
3673        Dulaglutide_Inj 1.5mg/0.5ml Pf Dev    1.363601        78
30        Abasaglar KwikPen_100u/ml 3ml Pf Pen    1.283053        55

In [215]:
final = pd.concat([drugs.iloc[0:50], drugs.iloc[len(drugs)-50: len(drugs)]], axis=0)
final.head()

bnf_name  rel_growth  count_16
1483          Butec_Transdermal Patch 10mcg/hr    3.946340        57
1485           Butec_Transdermal Patch 5mcg/hr    2.672189        75
4483  Fostair NEXThaler_Inh 200mcg/6mcg (120D)    1.726507        77
3673        Dulaglutide_Inj 1.5mg/0.5ml Pf Dev    1.363601        78
30        Abasaglar KwikPen_100u/ml 3ml Pf Pen    1.283053        55

In [216]:
arr = final.values
arr[:5]

array([['Butec_Transdermal Patch 10mcg/hr', 3.946340409455863, 57],
       ['Butec_Transdermal Patch 5mcg/hr', 2.672188773229457, 75],
       ['Fostair NEXThaler_Inh 200mcg/6mcg (120D)', 1.7265072272651727, 77],
       ['Dulaglutide_Inj 1.5mg/0.5ml Pf Dev', 1.363600606294229, 78],
       ['Abasaglar KwikPen_100u/ml 3ml Pf Pen', 1.2830530392006783, 55]], dtype=object)

In [217]:
lst=[]

for item in arr:
    lst.append(tuple(item))

In [218]:
def script_growth():
    return lst

In [219]:
grader.score('dw__script_growth', script_growth)

Your score:  1.0


## Question 6: rare_scripts

Does a practice's prescription costs originate from routine care or from reliance on rarely prescribed treatments? Commonplace treatments can carry lower costs than rare treatments because of efficiencies in large-scale production. While some specialist practices can't help but avoid prescribing rare medicines because there are no alternatives, some practices may be prescribing a unnecessary amount of brand-name products when generics are available. Let's identify practices whose costs disproportionately originate from rarely prescribed items.

First we have to identify which `'bnf_code'` are rare. To do this, find the probability $p$ of a prescription having a particular `'bnf_code'` if the `'bnf_code'` was randomly chosen from the unique options in the beneficiary data. We will call a `'bnf_code'` rare if it is prescribed at a rate less than $0.1p$.

In [ ]:
p = ...
rates = ...
rare_codes = ...
scripts['rare'] = ...

In [27]:
# Calculating probability for each bnf_code
p = 1 / float(scripts['bnf_code'].nunique())

In [28]:
# Calculating rare prescription i.e ratio of count of bnf_code per prescription and total count

rates = scripts.groupby('bnf_code')['bnf_code'].count().rename('count_prescription').reset_index()
rates.head()

bnf_code  count_prescription
0  0101010C0                  30
1  0101010F0                   3
2  0101010G0                 364
3  0101010I0                  21
4  0101010J0                  38

In [29]:
total_count = scripts['bnf_code'].count().astype(float)
total_count

1001634.0

In [30]:
rare = rates['count_prescription'] / total_count
rare.head()

0    0.000030
1    0.000003
2    0.000363
3    0.000021
4    0.000038
Name: count_prescription, dtype: float64

In [31]:
#Filtering the records on the basis of  rate < 0.1p

rare_codes =  rates[rare < 0.1*p]['bnf_code'].tolist()
rare_codes[:5]

['0101010C0', '0101010F0', '0101010I0', '0101010J0', '0101010N0']

In [32]:
#Creating 'rare' column in scripts
scrpt = scripts
scrpt['rare'] = scrpt['bnf_code'].isin(rare_codes)

In [33]:
scrpt.head()

practice   bnf_code                             bnf_name  items     nic  \
0   B87016  213200001  MucoClear Sod Chlor 6% Inh Soln 4ml      1   38.94   
1   L85017  0503021C0                  Aciclovir_Tab 800mg      4   13.44   
2   M81001  090900000    SMA_High Energy Milk Ready To Use      1  147.60   
3   F85063  0601011L0          Ins Humalog_100u/ml 10ml Vl      1   33.22   
4   B86667  1001010P0               Naproxen_Tab E/C 500mg      1    3.51   

   act_cost  quantity   rare  
0     36.06        60  False  
1     12.49       140  False  
2    136.65     15000  False  
3     30.77         2  False  
4      3.36        28  False

In [34]:
# Calculate total sum of act_cost for all treatment per prescription
treatment_sum = scrpt.groupby('practice')['act_cost'].sum()
treatment_sum.head()

practice
A81001     5447.55
A81002    42759.63
A81004    11804.14
A81005     9746.88
A81006    24645.30
Name: act_cost, dtype: float64

In [35]:
# Calculate cost for rare treatment per prescription
rare_treatment_sum = scrpt[scrpt['rare']].groupby('practice')['act_cost'].sum()
rare_treatment_sum.head()

practice
A81001     51.84
A81002    237.75
A81004      8.64
A81005    463.32
A81006     68.48
Name: act_cost, dtype: float64

In [121]:
# Calculate proportion of costs that originate from rare treatment i.e ratio of rare_treatment_sum and treatment_sum

rare_cost_prop = (rare_treatment_sum / treatment_sum)
rare_cost_prop.fillna(0, inplace=True)

mean_val = scrpt[scrpt['rare']]['act_cost'].sum() / scrpt['act_cost'].sum()

relative_rare_cost_prop = rare_cost_prop - mean_val 
standard_errors = relative_rare_cost_prop.std()

mean_val, standard_errors

(0.01554553433731762, 0.0678173558942156)

In [122]:
rare_scores = (relative_rare_cost_prop / standard_errors).rename('z-score').reset_index()
rare_scores.columns =['code','z-score']
rare_scores.head()

code   z-score
0  A81001 -0.088905
1  A81002 -0.147239
2  A81004 -0.218434
3  A81005  0.471703
4  A81006 -0.188254

In [38]:
# Sorting rare_score in descendig order of z-score
rare_scores.sort_values('z-score', ascending=False, inplace=True)
rare_scores.columns =['code','z-score']
rare_scores.head()

code    z-score
8034  Y03106  14.516261
8760  Y04676  14.516261
7441  Y00799  14.516261
8403  Y03976  14.516261
7600  Y01731  14.516261

In [88]:
practices.head()

code                        name                   addr_1         addr_2  \
0  A81001         THE DENSHAM SURGERY        THE HEALTH CENTRE  LAWSON STREET   
1  A81002  QUEENS PARK MEDICAL CENTRE  QUEENS PARK MEDICAL CTR  FARRER STREET   
2  A81003   VICTORIA MEDICAL PRACTICE        THE HEALTH CENTRE  VICTORIA ROAD   
3  A81004      WOODLANDS ROAD SURGERY         6 WOODLANDS ROAD            NaN   
4  A81005          SPRINGWOOD SURGERY       SPRINGWOOD SURGERY   RECTORY LANE   

            borough    village post_code  
0  STOCKTON ON TEES  CLEVELAND  TS18 1HU  
1  STOCKTON ON TEES  CLEVELAND  TS18 2AW  
2        HARTLEPOOL  CLEVELAND  TS26 8DB  
3     MIDDLESBROUGH  CLEVELAND   TS1 3BE  
4       GUISBOROUGH        NaN  TS14 7DJ

In [89]:
pract = practices[['code','name']].drop_duplicates('code')
pract.head()

code                        name
0  A81001         THE DENSHAM SURGERY
1  A81002  QUEENS PARK MEDICAL CENTRE
2  A81003   VICTORIA MEDICAL PRACTICE
3  A81004      WOODLANDS ROAD SURGERY
4  A81005          SPRINGWOOD SURGERY

In [90]:
pract.shape

(10843, 2)

In [91]:
rare_scores.shape

(9230, 2)

In [92]:
#Joining rare_scores with practices to get (practice, name, z-score)
joined_data = pd.merge(rare_scores, pract, on='code', how ='inner')#.drop_duplicates(['code','name'])
joined_data.head()

code   z-score                             name
0  A81001 -0.088905              THE DENSHAM SURGERY
1  A81002 -0.147239       QUEENS PARK MEDICAL CENTRE
2  A81004 -0.218434           WOODLANDS ROAD SURGERY
3  A81005  0.471703               SPRINGWOOD SURGERY
4  A81006 -0.188254  TENNANT STREET MEDICAL PRACTICE

In [103]:
joined_data.shape

(9230, 3)

In [108]:
joined_data.sort_values('z-score', ascending=False, inplace=True)

In [109]:
joined_data.reset_index(inplace=True)

In [110]:
required_cols = ['code', 'name', 'z-score']
final_df = joined_data[required_cols]

In [111]:
final_df.head()

code                           name    z-score
0  Y03106                      CRI GPWSI  14.516261
1  Y04704       PCOC RESPIRATORY SERVICE  14.516261
2  Y00215     ORTHOPAEDIC & RHEUMATOLOGY  14.516261
3  Y03268   CLECKHEATON DERMATOLOGY GPSI  14.516261
4  Y03572  DR WRIGHT ED CLINIC (BELMONT)  14.516261

In [114]:
final_df.shape

(9230, 3)

In [115]:
final_df = final_df.head(100)
arr = final_df.values

In [116]:
lst=[]

for item in arr:
    lst.append(tuple(item))

In [117]:
def rare_scripts():
    return lst

In [112]:
def rare_scripts():
    #return [("Y03106", "CRI GPWSI", 14.516)] * 100
    return [tuple(i) for i in final_df.values][:100]

Now for each practice, calculate the proportion of costs that originate from prescription of rare treatments (i.e. rare `'bnf_code'`). Use the `'act_cost'` field for this calculation.

In [ ]:
rare_cost_prop = ...

Now we will calculate a z-score for each practice based on this proportion.
First take the difference of `rare_cost_prop` and the proportion of costs originating from rare treatments across all practices.

In [ ]:
relative_rare_cost_prop = ...

Now we will estimate the standard errors (i.e. the denominator of the z-score) by simply taking the standard deviation of this difference.

In [ ]:
standard_errors = ...

Finally compute the z-scores. Return the practices with the top 100 z-scores in the form `(practice_name, proportion)`. Note that some practice codes will correspond with multiple names. In this case, use the first match when sorting names alphabetically.

In [ ]:
rare_scores = ...

In [165]:
def rare_scripts():
    return [("Y03106", "CRI GPWSI", 14.516)] * 100
    rare_scores.sort_values('act_cost', ascending=False, inplace=True)

In [118]:
grader.score('dw__rare_scripts', rare_scripts)

Your score:  0.96


*Copyright &copy; 2017 The Data Incubator.  All rights reserved.*